src/lib/completion.ts

```diff
@@ -1,131 +1,216 @@
+		this.summaryBot = new OpenAI({
+			openAIApiKey: openai_api_key,
+			modelName: this.model_name,
+			temperature: 0.0
 		});
-
-		this.qdrant_collection = qdrant_collection;
+		this.do_summaryBot = do_summaryBot;
+		this.qdrantClient = new QdrantClient({ host: 'localhost', port: 6333 });
+		this.collection_name = collection_name;
+		this.executor = undefined;
+		this.history = [];
+		this.do_history = do_history;
+		this.generate_data = generate_data;
+		this.chain = [];
+		if (openai_api_key == undefined) {
+			throw console.warn('OPENAI_API_KEY is undefined');
+		}
 	}
 
-	/*
-        We need to pass with_vector to qdrant to get our response
-    */
-	private async qdrant_similarity_search(query: string, k: number): Promise<Document[]> {
-		const query_embedding = await this.embeddings_model.embedQuery(query);
-		const qdrant_results = await this.qdrant_client.search(this.qdrant_collection, {
-			vector: query_embedding,
-			limit: k,
-			with_vector: true
+	public async setup() {
+		console.log(colors.fg.cyan, 'Model name: ' + this.model_name, colors.style.reset);
+		console.log('\n');
+		const tools = getTools(
+			this.qdrantClient,
+			this.collection_name,
+			this.embeddings_model,
+			this.summaryBot,
+			this.do_summaryBot
+		);
+		const model = new ChatOpenAI({
+			openAIApiKey: this.openai_api_key,
+			modelName: this.model_name,
+			temperature: 0.0,
+			stop: ['\nObservation']
 		});
-
-		console.log(qdrant_results);
-
-		// eslint-disable-next-line @typescript-eslint/no-explicit-any
-		return qdrant_results.map((result: any) => {
-			return new Document({
-				pageContent: result.payload.pageContent,
-				metadata: result.payload.metadata
-			});
+		const runnable = RunnableSequence.from([
+			{
+				input: (values: InputValues) => {
+					return values.input;
+				},
+				intermediate_steps: (values: InputValues) => {
+					return values.steps;
+				}
+			},
+			this.formatMessages,
+			model,
+			this.customOutputParser.bind(this)
+		]);
+		const executor = new AgentExecutor({
+			agent: runnable,
+			tools: tools
 		});
+		this.executor = executor;
 	}
-
-	private async get_vector_response(query: string): Promise<string[]> {
-		console.log('Retrieving vector response from qdrant...');
-
-		const vector_response = await this.qdrant_similarity_search(query, 2);
-
-		console.log(vector_response);
-
-		console.log('Vector response retreived');
-
-		// eslint-disable-next-line @typescript-eslint/no-explicit-any
-		return vector_response.map((document: Document<Record<string, any>>) => {
-			return document.pageContent;
-		});
-	}
-
-	private generate_history(history: ChatHistory[]): BaseMessage[] {
-		return history.map((message: { content: string; type: ChatHistoryType }) => {
-			if (message.type == ChatHistoryType.AI) {
-				return new AIMessage({ content: message.content });
-			} else {
-				return new HumanMessage({ content: message.content });
+	private formatMessages = async (values: InputValues) => {
```

Comment:
Add a comment here explaining what this function does or find ways to clean up the code we're doing because as it is this function is pretty confusing at a glance.